# Escola de Dados - Python para inovação cívica
## Módulo 1: Por dentro do python

- [Aula 5](#Aula-5:-Pacotes)
- [Aula 6](#Aula-6:-Método-readcsv)
- [Aula 7](#Aula-7:-Carregando-tabelas)


### Aula 5: Pacotes

Instalaremos todos os pacotes que usaremos nesse módulo; Pandas, Matplotlib, Numpy, Jupiter Widgets e Pandas Profiling. Descomente as linhas abaixo (remova o # do início dos comandos) para instalar.

In [ ]:
# pip install pandas

In [ ]:
# pip install matplotlib

In [ ]:
# pip install numpy

In [ ]:
# pip install ipywidgets

In [ ]:
# pip install pandas-profiling

A instalação é só realizada 1 vez, mas a importação das bibliotecas deve ocorrer no inicío do código, como na próxima linha:

In [1]:
#importa o pacote pandas e define um apelido para o mesmo: pd
import pandas as pd

### Aula 6: Método readcsv
Vamos aprender a ler arquivos do tipo CSV e converter para dataframe Pandas.

Os dados que usaremos nesse notebook pode ser encontrado em: [TSE. Candidaturas em 2020](https://www.tse.jus.br/eleicoes/estatisticas/repositorio-de-dados-eleitorais-1)

In [2]:
# utiliza o método read_csv(<caminho>, <caractere separador>, <encoding>) para
# atribuir à variável 'candidaturas_pe' o arquivo .csv e transformar a variável em um DATAFRAME
candidaturas_pe = pd.read_csv('base_dados/consulta_cand_2020_PE.csv', sep=';', encoding='latin_1')

In [3]:
# o comando abaixo retira a limitação da quantidade de colunas impressa na tela
# ou seja, veremos todas as colunas que constam no dataframe
pd.options.display.max_columns = None

In [4]:
# o método (ou função) head() exibe, por padrão, as cinco primeiras linhas do dataframe
candidaturas_pe.head()

DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
0  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
1  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
2  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
3  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
4  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   

   NR_TURNO  CD_ELEICAO                DS_ELEICAO  DT_ELEICAO TP_ABRANGENCIA  \
0         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
1         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
2         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
3         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
4         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   

  SG_UF  SG_UE        NM_UE  CD_CARGO  DS_CARGO  SQ_CANDIDATO  NR_CANDIDATO  \
0    PE  23817      CARUARU        13  VEREADOR  170001231149         17008   
1    PE  23094  ÁGUAS BELAS        13  VEREADOR  170000800795         15515   
2    PE  25313       RECIFE        13  VEREADOR  170000836659         25033   
3    PE  24970     OURICURI        13  VEREADOR  170000648896         45777   
4    PE  24996     PALMARES        13  VEREADOR  170001129430         90090   

                        NM_CANDIDATO         NM_URNA_CANDIDATO  \
0    MARIA DE LOURDES ALVES DE SOUZA   DONA LOURDES DA BONANÇA   
1  JOSÉ MAREVALDO BARROS DE OLIVEIRA       MAREVALDO DA RUA 15   
2            JOSÉ AGNIS PERKLES RÊGO                JOSÉ AGNIS   
3               DELVANI SILVA SOBRAL           DELVANIA SOBRAL   
4       MARILENE SHEILLA DE OLIVEIRA  SHEILLA DO MEIO AMBIENTE   

  NM_SOCIAL_CANDIDATO  NR_CPF_CANDIDATO                     NM_EMAIL  \
0              #NULO#       12310328472   NG_CONSULTORIA@HOTMAIL.COM   
1              #NULO#       59971576449  MDBABCAMPANHA2020@GMAIL.COM   
2              #NULO#       61079570730  DIRETORIODEMREC25@GMAIL.COM   
3              #NULO#       68068123415  DELVANIA_SOBRAL@HOTMAIL.COM   
4              #NULO#        4522160445      MARC_GUEDES@HOTMAIL.COM   

   CD_SITUACAO_CANDIDATURA DS_SITUACAO_CANDIDATURA  CD_DETALHE_SITUACAO_CAND  \
0                       12                    APTO                         2   
1                       12                    APTO                         2   
2                       12                    APTO                         2   
3                       12                    APTO                         2   
4                        3                  INAPTO                         6   

  DS_DETALHE_SITUACAO_CAND    TP_AGREMIACAO  NR_PARTIDO SG_PARTIDO  \
0                 DEFERIDO  PARTIDO ISOLADO          17        PSL   
1                 DEFERIDO  PARTIDO ISOLADO          15        MDB   
2                 DEFERIDO  PARTIDO ISOLADO          25        DEM   
3                 DEFERIDO  PARTIDO ISOLADO          45       PSDB   
4                 RENÚNCIA  PARTIDO ISOLADO          90       PROS   

                                NM_PARTIDO  SQ_COLIGACAO     NM_COLIGACAO  \
0                   PARTIDO SOCIAL LIBERAL  170000152093  PARTIDO ISOLADO   
1         MOVIMENTO DEMOCRÁTICO BRASILEIRO  170000078841  PARTIDO ISOLADO   
2                               DEMOCRATAS  170000088761  PARTIDO ISOLADO   
3  PARTIDO DA SOCIAL DEMOCRACIA BRASILEIRA  170000057628  PARTIDO ISOLADO   
4      PARTIDO REPUBLICANO DA ORDEM SOCIAL  170000136900  PARTIDO ISOLADO   

  DS_COMPOSICAO_COLIGACAO  CD_NACIONALIDADE           DS_NACIONALIDADE  \
0                     PSL                 1            BRASILEIRA NATA   
1                     MDB                 1            BRASILEIRA NATA   
2                     DEM                 1            BRASILEIRA NATA   
3                    PSDB                 2  BRASILEIRA (NATURALIZADA)   
4                  

In [5]:
# dtypes é uma função do pandas que retorna os tipos de dados que compoem determinado dataframe
candidaturas_pe.dtypes

DT_GERACAO                      object
HH_GERACAO                      object
ANO_ELEICAO                      int64
CD_TIPO_ELEICAO                  int64
NM_TIPO_ELEICAO                 object
                                 ...  
CD_SITUACAO_CANDIDATO_PLEITO     int64
DS_SITUACAO_CANDIDATO_PLEITO    object
CD_SITUACAO_CANDIDATO_URNA       int64
DS_SITUACAO_CANDIDATO_URNA      object
ST_CANDIDATO_INSERIDO_URNA      object
Length: 63, dtype: object

In [6]:
# já sabemos que o método dataframe.dtypes é utilizado para verificar os tipos de dados de cada coluna do dataframe
# porém, como o dataframe em questão possui diversas linhas e colunas, para visualizar os tipos de dados de todas
# as colunas, vamos verificar os tipos de dados em formato de dicionário
dict(candidaturas_pe.dtypes)

{'DT_GERACAO': dtype('O'),
 'HH_GERACAO': dtype('O'),
 'ANO_ELEICAO': dtype('int64'),
 'CD_TIPO_ELEICAO': dtype('int64'),
 'NM_TIPO_ELEICAO': dtype('O'),
 'NR_TURNO': dtype('int64'),
 'CD_ELEICAO': dtype('int64'),
 'DS_ELEICAO': dtype('O'),
 'DT_ELEICAO': dtype('O'),
 'TP_ABRANGENCIA': dtype('O'),
 'SG_UF': dtype('O'),
 'SG_UE': dtype('int64'),
 'NM_UE': dtype('O'),
 'CD_CARGO': dtype('int64'),
 'DS_CARGO': dtype('O'),
 'SQ_CANDIDATO': dtype('int64'),
 'NR_CANDIDATO': dtype('int64'),
 'NM_CANDIDATO': dtype('O'),
 'NM_URNA_CANDIDATO': dtype('O'),
 'NM_SOCIAL_CANDIDATO': dtype('O'),
 'NR_CPF_CANDIDATO': dtype('int64'),
 'NM_EMAIL': dtype('O'),
 'CD_SITUACAO_CANDIDATURA': dtype('int64'),
 'DS_SITUACAO_CANDIDATURA': dtype('O'),
 'CD_DETALHE_SITUACAO_CAND': dtype('int64'),
 'DS_DETALHE_SITUACAO_CAND': dtype('O'),
 'TP_AGREMIACAO': dtype('O'),
 'NR_PARTIDO': dtype('int64'),
 'SG_PARTIDO': dtype('O'),
 'NM_PARTIDO': dtype('O'),
 'SQ_COLIGACAO': dtype('int64'),
 'NM_COLIGACAO': dtype('O'),
 'D

In [8]:
# head() pode exibir quantas linhas o usuário desejar. Basta informar um parâmetro para o método.
# para visualizar as 10 primeiras linhas, pode-se usar head(10)
candidaturas_pe.head(10)

DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
0  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
1  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
2  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
3  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
4  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
5  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
6  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
7  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
8  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
9  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   

   NR_TURNO  CD_ELEICAO                DS_ELEICAO  DT_ELEICAO TP_ABRANGENCIA  \
0         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
1         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
2         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
3         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
4         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
5         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
6         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
7         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
8         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   
9         1         426  Eleições Municipais 2020  15/11/2020      MUNICIPAL   

  SG_UF  SG_UE        NM_UE  CD_CARGO  DS_CARGO  SQ_CANDIDATO  NR_CANDIDATO  \
0    PE  23817      CARUARU        13  VEREADOR  170001231149         17008   
1    PE  23094  ÁGUAS BELAS        13  VEREADOR  170000800795         15515   
2    PE  25313       RECIFE        13  VEREADOR  170000836659         25033   
3    PE  24970     OURICURI        13  VEREADOR  170000648896         45777   
4    PE  24996     PALMARES        13  VEREADOR  170001129430         90090   
5    PE  24910       OLINDA        13  VEREADOR  170000933978         10555   
6    PE  25119     PAUDALHO        13  VEREADOR  170001251698         40123   
7    PE  23108   SANTA CRUZ        13  VEREADOR  170000642383         70444   
8    PE  25313       RECIFE        13  VEREADOR  170000836614         25604   
9    PE  25615  SÃO CAITANO        13  VEREADOR  170000826892         10611   

                         NM_CANDIDATO         NM_URNA_CANDIDATO  \
0     MARIA DE LOURDES ALVES DE SOUZA   DONA LOURDES DA BONANÇA   
1   JOSÉ MAREVALDO BARROS DE OLIVEIRA       MAREVALDO DA RUA 15   
2             JOSÉ AGNIS PERKLES RÊGO                JOSÉ AGNIS   
3                DELVANI SILVA SOBRAL           DELVANIA SOBRAL   
4        MARILENE SHEILLA DE OLIVEIRA  SHEILLA DO MEIO AMBIENTE   
5       SILVANIA MARIA DE MELO CABRAL           SILVANIA CABRAL   
6         LEANDRO TEIXEIRA DOS SANTOS   DOUTOR LEANDRO TEIXEIRA   
7  ANTONIO JOANDSON DE ALENCAR AMORIM                      PEPI   
8         MARCIA MARIA ALVES DA SILVA              MARCIA SILVA   
9            GEREMIAS MOISES DA SILVA           GERA EQUIPADORA   

  NM_SOCIAL_CANDIDATO  NR_CPF_CANDIDATO                         NM_EMAIL  \
0              #NULO#       12310328472       NG_CONSULTORIA@HOTMAIL.COM   
1              #NULO#       59971576449      MDBABCAMPANHA2020@GMAIL.COM   
2              #NULO#       61079570730      DIRETORIODEMREC25@GMAIL.COM   
3              #NULO#       68068123415      DELVANIA_SOBRAL@HOTMAIL.COM   
4              #NULO#        4522160445          MARC_GUEDES@HOTMAIL.COM   
5              #NULO#       32829124472      SILVANIAMCABRAL@OUTLOOK.COM   
6              #NULO#        7186253407       TEIXEIRASANTOSPE@GMAIL.COM   
7              #NULO#       12002776431             JOANDS

In [9]:
# analogamente ao método head(), temos o método tail(), que exibe as últimas cinco linhas do dataframe
candidaturas_pe.tail()

DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
21096  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21097  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21098  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21099  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21100  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   

       NR_TURNO  CD_ELEICAO                DS_ELEICAO  DT_ELEICAO  \
21096         1         426  Eleições Municipais 2020  15/11/2020   
21097         1         426  Eleições Municipais 2020  15/11/2020   
21098         1         426  Eleições Municipais 2020  15/11/2020   
21099         1         426  Eleições Municipais 2020  15/11/2020   
21100         1         426  Eleições Municipais 2020  15/11/2020   

      TP_ABRANGENCIA SG_UF  SG_UE          NM_UE  CD_CARGO  DS_CARGO  \
21096      MUNICIPAL    PE  24716  LAGOA DO OURO        13  VEREADOR   
21097      MUNICIPAL    PE  23914        CONDADO        13  VEREADOR   
21098      MUNICIPAL    PE  23710      CAMUTANGA        13  VEREADOR   
21099      MUNICIPAL    PE  26310   ABREU E LIMA        13  VEREADOR   
21100      MUNICIPAL    PE  24996       PALMARES        13  VEREADOR   

       SQ_CANDIDATO  NR_CANDIDATO                     NM_CANDIDATO  \
21096  170000831074         11333  ADERBAL QUEIROZ MONTEIRO JUNIOR   
21097  170000822680         14114        BERNADETE MONTEIRO FALCÃO   
21098  170000758162         36555         GRAZIELE ROCHA DAMASCENO   
21099  170000782931         23232     CLAUDENIER ALMEIDA DE ARAÚJO   
21100  170001191022         45190          ELIAS SILVA DE OLIVEIRA   

                    NM_URNA_CANDIDATO NM_SOCIAL_CANDIDATO  NR_CPF_CANDIDATO  \
21096                 JÚNIOR MONTEIRO              #NULO#       36067776472   
21097                BERNADETE FALCÃO              #NULO#       19696027400   
21098                           GRAZI              #NULO#       10361236778   
21099  NIER ARAÚJO MANDATO COLABORATI              #NULO#       84571489404   
21100                    ELIAS DO GÁS              #NULO#        2699308412   

                          NM_EMAIL  CD_SITUACAO_CANDIDATURA  \
21096     PAULINHOPT13@HOTMAIL.COM                       12   
21097  BERNADETEFALCAO@HOTMAIL.COM                       12   
21098       PTCCAMUTANGA@GMAIL.COM                       12   
21099             NIER@NIER.COM.BR                       12   
21100      MARC_GUEDES@HOTMAIL.COM                       12   

      DS_SITUACAO_CANDIDATURA  CD_DETALHE_SITUACAO_CAND  \
21096                    APTO                         2   
21097                    APTO                         2   
21098                    APTO                         2   
21099                    APTO                         2   
21100                    APTO                         2   

      DS_DETALHE_SITUACAO_CAND    TP_AGREMIACAO  NR_PARTIDO SG_PARTIDO  \
21096                 DEFERIDO  PARTIDO ISOLADO          11         PP   
21097                 DEFERIDO  PARTIDO ISOLADO          14        PTB   
21098                 DEFERIDO  PARTIDO ISOLADO          36        PTC   
21099                 DEFERIDO  PARTIDO ISOLADO          23  CIDADANIA   
21100                 DEFERIDO  PARTIDO ISOLADO          45       PSDB   

                                    NM_PARTIDO  SQ_COLIGACAO     NM_COLIGACAO  \
21096                            PROGRESSISTAS  170000087696  PARTIDO ISOLADO   
21097           PARTIDO TRABALHISTA BRASILEIRO  170000086146  PARTIDO ISOLADO   
21098              PARTIDO TRABALHISTA CRISTÃO  170000072791  PARTIDO ISOLADO   
21099                                CIDADANIA  170000076102  PARTIDO ISOLADO   
21100  PARTIDO DA SOCIAL DEMOCRACIA BRASILEIRA  170000146430  PARTIDO ISOLADO   

      DS_COMPOSICAO_COLIGACAO  CD_NACIONALIDADE DS_NACIONALIDADE  \
21096                      PP             

In [10]:
# o método tail() também pode ser parametrizado para exibir a quantidade de linhas desejada pelo usuário
candidaturas_pe.tail(10)

DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
21091  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21092  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21093  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21094  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21095  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21096  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21097  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21098  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21099  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21100  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   

       NR_TURNO  CD_ELEICAO                DS_ELEICAO  DT_ELEICAO  \
21091         1         426  Eleições Municipais 2020  15/11/2020   
21092         1         426  Eleições Municipais 2020  15/11/2020   
21093         1         426  Eleições Municipais 2020  15/11/2020   
21094         1         426  Eleições Municipais 2020  15/11/2020   
21095         1         426  Eleições Municipais 2020  15/11/2020   
21096         1         426  Eleições Municipais 2020  15/11/2020   
21097         1         426  Eleições Municipais 2020  15/11/2020   
21098         1         426  Eleições Municipais 2020  15/11/2020   
21099         1         426  Eleições Municipais 2020  15/11/2020   
21100         1         426  Eleições Municipais 2020  15/11/2020   

      TP_ABRANGENCIA SG_UF  SG_UE                     NM_UE  CD_CARGO  \
21091      MUNICIPAL    PE  23973                    CUMARU        13   
21092      MUNICIPAL    PE  23515    BREJO DA MADRE DE DEUS        12   
21093      MUNICIPAL    PE  24910                    OLINDA        13   
21094      MUNICIPAL    PE  24759                    LAJEDO        13   
21095      MUNICIPAL    PE  25496  SANTA CRUZ DO CAPIBARIBE        13   
21096      MUNICIPAL    PE  24716             LAGOA DO OURO        13   
21097      MUNICIPAL    PE  23914                   CONDADO        13   
21098      MUNICIPAL    PE  23710                 CAMUTANGA        13   
21099      MUNICIPAL    PE  26310              ABREU E LIMA        13   
21100      MUNICIPAL    PE  24996                  PALMARES        13   

            DS_CARGO  SQ_CANDIDATO  NR_CANDIDATO  \
21091       VEREADOR  170001220312         55369   
21092  VICE-PREFEITO  170001145119            22   
21093       VEREADOR  170000850458         55655   
21094       VEREADOR  170001036680         36123   
21095       VEREADOR  170000998830         28123   
21096       VEREADOR  170000831074         11333   
21097       VEREADOR  170000822680         14114   
21098       VEREADOR  170000758162         36555   
21099       VEREADOR  170000782931         23232   
21100       VEREADOR  170001191022         45190   

                            NM_CANDIDATO               NM_URNA_CANDIDATO  \
21091              ADEILDA LÚCIA PEREIRA                 ADEILDA PEREIRA   
21092            RUBIENO MARQUES DE MELO                 RUBIENO CATANHA   
21093    JOSÉ FLAVIO ALVES DO NASCIMENTO               FLAVIO NASCIMENTO   
21094          ALEXSANDRA MARIA DA SILVA                      ALEXSANDRA   
21095  ISABEL CRISTINA TEIXEIRA DA SILVA                ISABEL DE VALDIR   
21096    ADERBAL QUEIROZ MONTEIRO JUNIOR                 JÚNIOR MONTEIRO   
21097          BERNADETE MONTEIRO FALCÃO                BERNADETE FALCÃO   
21098           GRAZIELE ROCHA DAMASCENO                           GRAZI   
21099       CLAUDENIER ALMEIDA DE ARAÚJO  NIER ARAÚJO MANDATO COLABORATI   
21100            ELIAS SILVA DE OLIVEIRA                    ELIAS DO GÁS   

      NM_SOCIAL_CANDIDATO  NR_CPF_CANDIDATO                        NM_EMAIL  \
21091              #NULO#       35072385434      UNIAOPORCUMAR

In [11]:
# o método describe() do pandas gera estatísticas descritivas do dataframe
candidaturas_pe.describe()

ANO_ELEICAO  CD_TIPO_ELEICAO      NR_TURNO    CD_ELEICAO         SG_UE  \
count      21101.0          21101.0  21101.000000  21101.000000  21101.000000   
mean        2020.0              2.0      1.000379    426.000379  24650.892943   
std            0.0              0.0      0.019468      0.019468    956.269929   
min         2020.0              2.0      1.000000    426.000000  23000.000000   
25%         2020.0              2.0      1.000000    426.000000  23817.000000   
50%         2020.0              2.0      1.000000    426.000000  24732.000000   
75%         2020.0              2.0      1.000000    426.000000  25313.000000   
max         2020.0              2.0      2.000000    427.000000  26336.000000   

           CD_CARGO  SQ_CANDIDATO  NR_CANDIDATO  NR_CPF_CANDIDATO  \
count  21101.000000  2.110100e+04  21101.000000      2.110100e+04   
mean      12.905170  1.700010e+11  30485.193214      3.000134e+10   
std        0.385719  1.853569e+05  21909.512450      3.101523e+10   
min       11.000000  1.700006e+11     10.000000      1.321943e+07   
25%       13.000000  1.700008e+11  13789.000000      4.876538e+09   
50%       13.000000  1.700010e+11  22789.000000      1.106791e+10   
75%       13.000000  1.700011e+11  40999.000000      5.736897e+10   
max       13.000000  1.700013e+11  90999.000000      9.997776e+10   

       CD_SITUACAO_CANDIDATURA  CD_DETALHE_SITUACAO_CAND    NR_PARTIDO  \
count             21101.000000              21101.000000  21101.000000   
mean                 11.528695                  2.474006     32.259561   
std                   2.004945                  2.197985     21.119223   
min                   3.000000                  2.000000     10.000000   
25%                  12.000000                  2.000000     15.000000   
50%                  12.000000                  2.000000     25.000000   
75%                  12.000000                  2.000000     45.000000   
max                  12.000000                 16.000000     90.000000   

       SQ_COLIGACAO  CD_NACIONALIDADE  CD_MUNICIPIO_NASCIMENTO  \
count  2.110100e+04      21101.000000                  21101.0   
mean   1.700001e+11          1.005213                     -3.0   
std    3.073051e+04          0.073963                      0.0   
min    1.700001e+11          1.000000                     -3.0   
25%    1.700001e+11          1.000000                     -3.0   
50%    1.700001e+11          1.000000                     -3.0   
75%    1.700001e+11          1.000000                     -3.0   
max    1.700002e+11          4.000000                     -3.0   

       NR_IDADE_DATA_POSSE  NR_TITULO_ELEITORAL_CANDIDATO     CD_GENERO  \
count         21101.000000                   2.110100e+04  21101.000000   
mean             44.838491                   5.146977e+10      2.651154   
std              11.413648                   3.544932e+10      0.937202   
min              16.000000                   8.270833e+06      2.000000   
25%              37.000000                   3.288344e+10      2.000000   
50%              44.000000                   4.861352e+10      2.000000   
75%              53.000000                   6.508857e+10      4.000000   
max              93.000000                   4.444922e+11      4.000000   

       CD_GRAU_INSTRUCAO  CD_ESTADO_CIVIL   CD_COR_RACA   CD_OCUPACAO  \
count       21101.000000     21101.000000  21101.000000  21101.000000   
mean            5.643524         2.733851      2.213355    536.581821   
std             1.718737         2.058681      0.997568    327.416016   
min             1.000000         1.000000      1.000000    101.000000   
25%             4.000000         1.000000      1.000000    257.000000   
50%             6.000000         3.000000      3.000000    536.000000   
75%             6.000000         3.000000      3.000000    931.000000   
max             8.000000         9.000000      6.000000    999.000000   

       VR_DESPESA_MAX_CAMPANHA  CD_SIT_TOT_TURNO  NR_PROTOC

In [12]:
# o método sample() extrai uma linha de amostra do dataframe. Também pode ser parametrizado para informar n linhas
candidaturas_pe.sample(2)

DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
9430  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
7439  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   

      NR_TURNO  CD_ELEICAO                DS_ELEICAO  DT_ELEICAO  \
9430         1         426  Eleições Municipais 2020  15/11/2020   
7439         1         426  Eleições Municipais 2020  15/11/2020   

     TP_ABRANGENCIA SG_UF  SG_UE              NM_UE  CD_CARGO  DS_CARGO  \
9430      MUNICIPAL    PE  25178          PESQUEIRA        13  VEREADOR   
7439      MUNICIPAL    PE  24511  ILHA DE ITAMARACÁ        13  VEREADOR   

      SQ_CANDIDATO  NR_CANDIDATO                 NM_CANDIDATO  \
9430  170001059617         13000         EDSON GOMES DA SILVA   
7439  170001126575         11500  CLAUDEMIR PEDRO DE OLIVEIRA   

     NM_URNA_CANDIDATO NM_SOCIAL_CANDIDATO  NR_CPF_CANDIDATO  \
9430    DINHO BARBEIRO              #NULO#        2455141470   
7439      DEMA DO ALTO              #NULO#        2705787496   

                         NM_EMAIL  CD_SITUACAO_CANDIDATURA  \
9430  EDSONDASILVA.0102@GMAIL.COM                       12   
7439         MARRTINS06@GMAIL.COM                       12   

     DS_SITUACAO_CANDIDATURA  CD_DETALHE_SITUACAO_CAND  \
9430                    APTO                         2   
7439                    APTO                         2   

     DS_DETALHE_SITUACAO_CAND    TP_AGREMIACAO  NR_PARTIDO SG_PARTIDO  \
9430                 DEFERIDO  PARTIDO ISOLADO          13         PT   
7439                 DEFERIDO  PARTIDO ISOLADO          11         PP   

                     NM_PARTIDO  SQ_COLIGACAO     NM_COLIGACAO  \
9430  PARTIDO DOS TRABALHADORES  170000128120  PARTIDO ISOLADO   
7439              PROGRESSISTAS  170000136562  PARTIDO ISOLADO   

     DS_COMPOSICAO_COLIGACAO  CD_NACIONALIDADE DS_NACIONALIDADE  \
9430                      PT                 1  BRASILEIRA NATA   
7439                      PP                 1  BRASILEIRA NATA   

     SG_UF_NASCIMENTO  CD_MUNICIPIO_NASCIMENTO NM_MUNICIPIO_NASCIMENTO  \
9430               PE                       -3               PESQUEIRA   
7439               PE                       -3                IGARASSU   

     DT_NASCIMENTO  NR_IDADE_DATA_POSSE  NR_TITULO_ELEITORAL_CANDIDATO  \
9430    01/02/1978                   42                    49674130876   
7439    11/11/1962                   58                    23685360809   

      CD_GENERO  DS_GENERO  CD_GRAU_INSTRUCAO              DS_GRAU_INSTRUCAO  \
9430          2  MASCULINO                  2                   LÊ E ESCREVE   
7439          2  MASCULINO                  3  ENSINO FUNDAMENTAL INCOMPLETO   

      CD_ESTADO_CIVIL DS_ESTADO_CIVIL  CD_COR_RACA DS_COR_RACA  CD_OCUPACAO  \
9430                1     SOLTEIRO(A)            3       PARDA          999   
7439                3       CASADO(A)            2       PRETA          999   

     DS_OCUPACAO  VR_DESPESA_MAX_CAMPANHA  CD_SIT_TOT_TURNO DS_SIT_TOT_TURNO  \
9430      OUTROS                 23077.95                 4       NÃO ELEITO   
7439      OUTROS                 29696.26                 2    ELEITO POR QP   

     ST_REELEICAO ST_DECLARAR_BENS  NR_PROTOCOLO_CANDIDATURA  \
9430            N                N                        -1   
7439            N                S                        -1   

              NR_PROCESSO  CD_SITUACAO_CANDIDATO_PLEITO  \
9430  6001533520206170055                             2   
7439  6002353220206170131                             2   

     DS_SITUACAO_CANDIDATO_PLEITO  CD_SITUACAO_CANDIDATO_URNA  \
9430                     DEFERIDO                           2   
7439                     DEFERIDO                           4   

     DS_SITUACAO_CANDIDATO_URNA ST_CANDIDATO_INSERIDO_URNA  
9430                   DEFERIDO                        SIM  
7439     INDEFERIDO COM RECURSO                        SIM

In [13]:
#o atributo shape informa o tamanho do dataframe (quantidade de linhas e colunas)
candidaturas_pe.shape

(21101, 63)

## Quantas candidaturas houveram no seu município nas eleições de 2020?

#### Antes de responder essa pergunta, vamos verificar, primeiro, quais cidades existem no dataframe *candidaturas_pe*

In [14]:
# para responder a questão, vamos criar uma variável para armazenar os valores retornados pelo dataframe através do método 
# unique()
# o método unique() retorna valores únicos de um objeto do tipo Série, em formato de array
# uma coluna de um dataframe pode ser considerado um objeto do tipo Série
# por isso utilizamos a sintaxe candidaturas_pe['NM_UE'] para extair dela os valores únicos dessa série, ou seja,
# quais são as cidades que fazem parte do Estado de Pernambuco - 'NM_UE' significa NOME DA UNIDADE DO ESTADO
cidades_pe = candidaturas_pe['NM_UE'].unique()
cidades_pe

array(['CARUARU', 'ÁGUAS BELAS', 'RECIFE', 'OURICURI', 'PALMARES',
       'OLINDA', 'PAUDALHO', 'SANTA CRUZ', 'SÃO CAITANO', 'GOIANA',
       'PETROLÂNDIA', 'MORENO', 'LAGOA DE ITAENGA', 'PAULISTA',
       'ABREU E LIMA', 'BOM CONSELHO', 'GRAVATÁ', 'PASSIRA',
       'AFOGADOS DA INGAZEIRA', 'JAQUEIRA', 'ARCOVERDE', 'MARAIAL',
       'PARANATAMA', 'SIRINHAÉM', 'ITAPISSUMA', 'RIO FORMOSO',
       'JABOATÃO DOS GUARARAPES', 'IGARASSU', 'BUÍQUE', 'SALOÁ',
       'IBIRAJUBA', 'POMBOS', 'FLORESTA', 'SANTA MARIA DA BOA VISTA',
       'EXU', 'GLÓRIA DO GOITÁ', 'CARPINA', 'SURUBIM', 'LAGOA DO CARRO',
       'AGRESTINA', 'TRIUNFO', 'MACAPARANA', 'CUMARU', 'CONDADO',
       'CAMARAGIBE', 'CARNAÍBA', 'CABO DE SANTO AGOSTINHO', 'TABIRA',
       'TUPANATINGA', 'ALAGOINHA', 'CATENDE', 'ESCADA',
       'CAMOCIM DE SÃO FÉLIX', 'SALGUEIRO', 'TRINDADE', 'ARARIPINA',
       'ITAÍBA', 'PRIMAVERA', 'TRACUNHAÉM', 'BEZERROS', 'TUPARETAMA',
       'TACAIMBÓ', 'SANTA MARIA DO CAMBUCÁ', 'VITÓRIA DE SANTO ANTÃO',

In [15]:
# aqui, verificamos o tamanho da lista retornada pelo objeto Série do nosso dataframe
# através do método len(), obtemos o valor 184 como o número de cidades constantes no array "cidades_pe"
len(cidades_pe)

184

#### Por meio de pesquisas, descobri outro método para verificar valores únicos em uma Série. É através do método set(), que constrói um tipo de dados chamado CONJUNTO, no qual somente podem existir valores EXCLUSIVOS. Podemos utilizar o método set() para encontrar as mesmas informações que encontramos através de unique()

In [16]:
# transformando a variável "cidades" em um conjunto dos valores exclusivos da Série "NM_UE" do dataframe "candidaturas_pe"
cidades = set(candidaturas_pe['NM_UE'])

In [17]:
# exibindo as informações
cidades

{'ABREU E LIMA',
 'AFOGADOS DA INGAZEIRA',
 'AFRÂNIO',
 'AGRESTINA',
 'ALAGOINHA',
 'ALIANÇA',
 'ALTINHO',
 'AMARAJI',
 'ANGELIM',
 'ARARIPINA',
 'ARAÇOIABA',
 'ARCOVERDE',
 'BARRA DE GUABIRABA',
 'BARREIROS',
 'BELO JARDIM',
 'BELÉM DE MARIA',
 'BELÉM DO SÃO FRANCISCO',
 'BETÂNIA',
 'BEZERROS',
 'BODOCÓ',
 'BOM CONSELHO',
 'BOM JARDIM',
 'BONITO',
 'BREJINHO',
 'BREJO DA MADRE DE DEUS',
 'BREJÃO',
 'BUENOS AIRES',
 'BUÍQUE',
 'CABO DE SANTO AGOSTINHO',
 'CABROBÓ',
 'CACHOEIRINHA',
 'CAETÉS',
 'CALUMBI',
 'CALÇADO',
 'CAMARAGIBE',
 'CAMOCIM DE SÃO FÉLIX',
 'CAMUTANGA',
 'CANHOTINHO',
 'CAPOEIRAS',
 'CARNAUBEIRA DA PENHA',
 'CARNAÍBA',
 'CARPINA',
 'CARUARU',
 'CASINHAS',
 'CATENDE',
 'CEDRO',
 'CHÃ DE ALEGRIA',
 'CHÃ GRANDE',
 'CONDADO',
 'CORRENTES',
 'CORTÊS',
 'CUMARU',
 'CUPIRA',
 'CUSTÓDIA',
 'DORMENTES',
 'ESCADA',
 'EXU',
 'FEIRA NOVA',
 'FERREIROS',
 'FLORES',
 'FLORESTA',
 'FREI MIGUELINHO',
 'GAMELEIRA',
 'GARANHUNS',
 'GLÓRIA DO GOITÁ',
 'GOIANA',
 'GRANITO',
 'GRAVATÁ',
 'I

In [18]:
# contando o número de valores no conjunto
len(cidades)

184

In [19]:
# vamos sortear um número aleatório entre 0 e 183 para escolher uma cidade para encontrar os dados
# para isso, precisamos importar a bibliotecam random
import random

In [20]:
# sorteando um número aleatório
numero_aleatorio = random.randint(0, 183)
numero_aleatorio

180

In [21]:
# a cidade escolhida para ser trabalhada, portanto, é a de posição 134 da lista cidades_pe
cidade = cidades_pe[134]
cidade

'SANTA TEREZINHA'

## Quantas candidaturas houveram no município de Santa Terezinha - PE nas eleições de 2020?

### Método 1 - loop for

In [22]:
# para encontrar o número de candidaturas, podemos contar quantas vezes a cidade de Santa Terezinha aparece na lista
# para isso, criamos um loop que verifique todos os valores da Série candidaturas_pe['NM_UE'] e some a um contador
# todas as vezes que o valor SANTA TEREZINHA aparecer na lista
cont = 0
for candidatura in candidaturas_pe['NM_UE']:
    if candidatura == cidade:
        cont = cont + 1
print(f'O total de candidaturas na cidade de {cidade} foi de {cont}')

O total de candidaturas na cidade de SANTA TEREZINHA foi de 41


### Método 2 - criando um dataframe a partir de um filtro do dataframe original

In [23]:
# aqui, candidaturas_st será um novo dataframe, com os valores filtrados do dataframe original
# o filtro é criado a partir dos valores da coluna 'NM_UE' que se igualam a 'SANTA TEREZINHA'
# ou seja, pegamos o dataframe candidaturas_pe e filtramos da Série candidaturas_pe['NM_UE'] apenas os valores
# que sejam iguais (==) a 'SANTA TEREZINHA'
candidaturas_st = candidaturas_pe[candidaturas_pe['NM_UE'] == 'SANTA TEREZINHA']

In [24]:
# vamos utilizar o atributo shape para verificar o tamanho da série filtrada
candidaturas_st.shape

(41, 63)

In [25]:
# como o atributo shape retorna uma lista com o número de linhas e colunas do dataframe, sabemos que o total de linhas
# representa o total de candidaturas. Portanto, a posição 0 de shape[0] é o total de candidaturas da cidade de 
# Santa Terezinha
candidaturas_st.shape[0]

41

In [26]:
total_candidaturas_st = candidaturas_st.shape[0]
print(f'O total de candidaturas na cidade de {cidade} é foi de {total_candidaturas_st}')

O total de candidaturas na cidade de SANTA TEREZINHA é foi de 41


#### Se quisermos encontrar o total de candidaturas para a cidade de Caruaru, por exemplo, podemos utilizar os mesmos métodos

### Método filtro de dataframe

In [27]:
candidaturas_cr = candidaturas_pe[candidaturas_pe['NM_UE'] == 'CARUARU']
candidaturas_cr.shape

(480, 63)

### Método iterador for

In [28]:
nr_caruaru = 0
for cidade in candidaturas_pe['NM_UE']:
    if cidade == 'CARUARU':
        nr_caruaru = nr_caruaru + 1
nr_caruaru

480

#### Abaixo utilizamos o método value_counts, cuja função é retornar uma Série com a contagem de valores únicos de alguma série
#### Esse método proporciona uma forma rápida de avaliar quantas vezes determinada informação aparece no dataframe
#### Para os nossos propósitos, essa função pode informar o número de candidaturas para cada cidade do Estado de Pernambuco

In [29]:
# value_counts exibe a contagem dos valores únicos de uma série. A contagem é decrescente, e valores do tipo NA são excluídos
cand_por_cidade_pe = candidaturas_pe['NM_UE'].value_counts()
cand_por_cidade_pe

RECIFE                     922
JABOATÃO DOS GUARARAPES    780
OLINDA                     604
CABO DE SANTO AGOSTINHO    583
PAULISTA                   578
                          ... 
JUPI                        24
INGAZEIRA                   23
QUIXABA                     23
GRANITO                     21
SOLIDÃO                     20
Name: NM_UE, Length: 184, dtype: int64

In [30]:
# vamos apresentá-la, também, em formato de dicionário, pois assim podemos visualizar todos os valores
dict(cand_por_cidade_pe)

{'RECIFE': 922,
 'JABOATÃO DOS GUARARAPES': 780,
 'OLINDA': 604,
 'CABO DE SANTO AGOSTINHO': 583,
 'PAULISTA': 578,
 'CARUARU': 480,
 'CAMARAGIBE': 478,
 'PETROLINA': 387,
 'SÃO LOURENÇO DA MATA': 361,
 'IPOJUCA': 319,
 'GRAVATÁ': 282,
 'GARANHUNS': 281,
 'VITÓRIA DE SANTO ANTÃO': 270,
 'ABREU E LIMA': 259,
 'CARPINA': 257,
 'GOIANA': 243,
 'IGARASSU': 240,
 'MORENO': 236,
 'PALMARES': 230,
 'ESCADA': 228,
 'CATENDE': 212,
 'LIMOEIRO': 210,
 'PAUDALHO': 196,
 'PESQUEIRA': 190,
 'BELO JARDIM': 189,
 'SANTA CRUZ DO CAPIBARIBE': 188,
 'SÃO CAITANO': 186,
 'NAZARÉ DA MATA': 183,
 'SURUBIM': 171,
 'PETROLÂNDIA': 171,
 'SERRA TALHADA': 169,
 'BEZERROS': 165,
 'SIRINHAÉM': 164,
 'RIBEIRÃO': 159,
 'TIMBAÚBA': 158,
 'BARREIROS': 156,
 'ARCOVERDE': 155,
 'OURICURI': 154,
 'ILHA DE ITAMARACÁ': 152,
 'SALGUEIRO': 148,
 'TAMANDARÉ': 142,
 'SÃO JOSÉ DA COROA GRANDE': 139,
 'PASSIRA': 137,
 'ÁGUAS BELAS': 135,
 'BONITO': 128,
 'RIO FORMOSO': 127,
 'SÃO BENTO DO UNA': 123,
 'LAGOA DO CARRO': 120,
 'BO

In [31]:
# também podemos apresentar os dados em formato de dataframe
cand_por_cidade_pe.to_frame()

NM_UE
RECIFE                     922
JABOATÃO DOS GUARARAPES    780
OLINDA                     604
CABO DE SANTO AGOSTINHO    583
PAULISTA                   578
...                        ...
JUPI                        24
INGAZEIRA                   23
QUIXABA                     23
GRANITO                     21
SOLIDÃO                     20

[184 rows x 1 columns]

####  Vamos construir um pequeno sistema de busca, por meio de sorteio, das cidades de Pernambuco e seu respectivo número de candidaturas

In [37]:
# primeiro colocamos em um dicionário todas as cidades e seus respectivos valores
cidades_dict = dict(cand_por_cidade_pe)

In [39]:
# depois, determinamos um número aleatório, para sortear a cidade a ser buscada
num_sorteio = random.randint(0,len(cand_por_cidade_pe))
# com o número definido, utilizamos esse número para buscar a chave do dicionário, que é a cidade sorteada
cidade_sorteada = list(cidades_dict)[num_sorteio]
# com a cidade sorteada, pesquisamos seu valor informando o dicionário a sua chave
cand_cidade_sorteada = cidades_dict[cidade_sorteada]
# e mostramos em uma mensagem qual foi a cidade sorteada e o número de candidaturas
print(f'A cidade de {cidade_sorteada} teve {cand_cidade_sorteada} candidatos na eleição municipal de 2018')

A cidade de VERTENTE DO LÉRIO teve 54 candidatos na eleição municipal de 2018


#### Também podemos criar esse mesmo sistema de busca através de filtro no dataframe

In [46]:
# definimos o nome da cidade em uma variável (também poderíamos criar um sistema de geraçaõ de número aleatório)
st = 'SANTA TEREZINHA'

In [48]:
# e filtramos o dataframe a partir da variável
df_cidade_st = candidaturas_pe[candidaturas_pe['NM_UE'] == st]
df_cidade_st.shape

(41, 63)

## Aula 7: Carregando tabelas

É possível ler vários arquivos CSV e converter de uma só vez para dataframe Pandas. Para isso, além de usar as técnicas vistas nas aulas anteriores, importaremos a biblioteca glob.

#### A bilblioteca "glob" encontra todos os nomes de caminho(*pathnames*) que correspondem ao padrão informado. Essa biblioteca geralmente é utilizada para agrupar diversos arquivos em uma só variavel

In [49]:
# importando a biblioteca glob
import glob

#### Pelo visto existem dois modos de trabalhar com a  função principal da biblioteca glob.
#### O primeiro modo é criando um iterador através do método "iglob". Esse método retorna uma lista de todos os caminhos que correspondem ao padrão informado, *porém sem armazená-los ao mesmo tempo* (ao contrário do que faz o método "glob", como veremos adiante).

#### Vamos verificar como funciona o método "iglob"

In [50]:
# declarando uma lista para armazenar todos os caminhos
df = []

In [51]:
# criando um loop for com o iterador retornado pelo método iglob para armaenzar na variável todos os caminhos
# encontrados pelo método
for dado in glob.iglob('base_dados/consulta_*.csv'):
    df_iglob  = pd.read_csv(dado, sep=';', encoding='latin_1')
    df.append(df_iglob)
df

[       DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
 0      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 1      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 2      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 3      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 4      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 ...           ...        ...          ...              ...                ...   
 21096  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21097  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21098  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21099  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21100  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 
        NR_TUR

In [52]:
df_2 = []
for dado in glob.glob('base_dados/consulta_*.csv'):
    df_glob = pd.read_csv(dado, sep=';', encoding='latin_1')
    df_2.append(df_glob)
df_2

[       DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
 0      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 1      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 2      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 3      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 4      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 ...           ...        ...          ...              ...                ...   
 21096  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21097  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21098  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21099  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21100  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 
        NR_TUR

#### Já o método "glob" *armazena* todos os caminhos *pathnames* que correspondem ao padrão informado. No exemplo abaixo, podemos visualizar o que está armazenado na variável "dados_todos_estados". Com o método "iglob", essa visualização seria impossível

In [53]:
# o método glob da biblioteca glob importa o caminho completo de todos os arquivos que correspondem ao padrão passado
# para ele. Portanto, a variável dados_todos_estados armazenará o caminho de todos os arquivos que estão em determinado
# local e possuem determinado nome
dados_todos_estados = glob.glob('base_dados/consulta_*.csv')
print(dados_todos_estados)

['base_dados\\consulta_cand_2020_PE.csv', 'base_dados\\consulta_cand_2020_RN.csv']


In [54]:
df_RN = pd.read_csv(dados_todos_estados[0], sep=";", encoding="latin_1")
df_RN.shape

(21101, 63)

In [120]:
df_RN = pd.read_csv('base_dados/consulta_cand_2020_RN.csv', sep=';', encoding='latin_1')

In [55]:
df_PE = pd.read_csv(dados_todos_estados[1], sep=';', encoding='latin_1')
df_PE.shape

(10555, 63)

In [56]:
# criamos uma variável para armazenar todos esses arquivos
lista_df = []

In [57]:
# antes de armazenar os arquivos, porém, precisamos transformá-los em dataframes
# o loop for abaixo irá entrar na lista onde estão armanzeados os caminhos e transformar cada um dos arquivos
# que constam lá em dataframes. Após isso, irá colocar cada dataframe na variável criada anteriormente, a lista_df

for dados_estado in dados_todos_estados:
    df_estado = pd.read_csv(dados_estado, sep=';', encoding='latin_1')
    lista_df.append(df_estado)
lista_df

[       DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
 0      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 1      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 2      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 3      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 4      05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 ...           ...        ...          ...              ...                ...   
 21096  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21097  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21098  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21099  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 21100  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
 
        NR_TUR

#### Vamos agora unir os dois dataframes. Faremos isso utilizando o método CONCAT

#### O método CONCAT concatena objetos panda *ao longo de um eixo* com lógica de conjunto opcional para os outros eixos

In [58]:
# agora, queremos que todos os arquivos sejam agrupados(concatenados) em um arquivo só
# para isso, utiliza-se o método CONCAT, que cria um dataframe único, unindo todos os dataframes
# que estavam individualizados na variável lista_df
df_completo = pd.concat(lista_df, axis=0, ignore_index=True)
# sobre os parâmetros do método concat:
# axis=0: significa "concatenar as linhas (index)"; com "axis=1" significaria "concatenar colunas"
# ignore_index=True: pede para que o dataframe final, concatenado, não considere os índices dos dataframes originais
# com ignore_index=False, teríamos os índices dos dataframos originas no dataframe concatenado
df_completo.shape

(31656, 63)

### Analisando o Dataframe

#### Vamos verificar quantos Estados (UF) existem no dataframe

In [59]:
# verificando quantos estados existem no dataframe por meio do método unique()
lista_uf = df_completo['SG_UF'].unique()
lista_uf

array(['PE', 'RN'], dtype=object)

#### Como visto anteriormente, podemos verificar os valores únicos de uma série através do construtor set()

In [60]:
set(df_completo['SG_UF'])

{'PE', 'RN'}

#### Verificamos que existem dois Estados nesse dataframe. Agora vamos encontrar o número de candidaturas para duas cidades de cada Estado

In [61]:
# primeiro vamos verificar quantas cidades temos para cada Estado
# colocando em uma lista as ocorrências de cidades no Estado de Pernambuco
total_cidades_pe = df_completo[df_completo['SG_UF'] == 'PE']

In [62]:
# vamos adicionar a uma lista os valores únicos de cidades no Estado de Pernambuco
cidades_pe = total_cidades_pe['NM_UE'].unique()
cidades_pe

array(['CARUARU', 'ÁGUAS BELAS', 'RECIFE', 'OURICURI', 'PALMARES',
       'OLINDA', 'PAUDALHO', 'SANTA CRUZ', 'SÃO CAITANO', 'GOIANA',
       'PETROLÂNDIA', 'MORENO', 'LAGOA DE ITAENGA', 'PAULISTA',
       'ABREU E LIMA', 'BOM CONSELHO', 'GRAVATÁ', 'PASSIRA',
       'AFOGADOS DA INGAZEIRA', 'JAQUEIRA', 'ARCOVERDE', 'MARAIAL',
       'PARANATAMA', 'SIRINHAÉM', 'ITAPISSUMA', 'RIO FORMOSO',
       'JABOATÃO DOS GUARARAPES', 'IGARASSU', 'BUÍQUE', 'SALOÁ',
       'IBIRAJUBA', 'POMBOS', 'FLORESTA', 'SANTA MARIA DA BOA VISTA',
       'EXU', 'GLÓRIA DO GOITÁ', 'CARPINA', 'SURUBIM', 'LAGOA DO CARRO',
       'AGRESTINA', 'TRIUNFO', 'MACAPARANA', 'CUMARU', 'CONDADO',
       'CAMARAGIBE', 'CARNAÍBA', 'CABO DE SANTO AGOSTINHO', 'TABIRA',
       'TUPANATINGA', 'ALAGOINHA', 'CATENDE', 'ESCADA',
       'CAMOCIM DE SÃO FÉLIX', 'SALGUEIRO', 'TRINDADE', 'ARARIPINA',
       'ITAÍBA', 'PRIMAVERA', 'TRACUNHAÉM', 'BEZERROS', 'TUPARETAMA',
       'TACAIMBÓ', 'SANTA MARIA DO CAMBUCÁ', 'VITÓRIA DE SANTO ANTÃO',

#### Utlizando sorte_values() para ordenar os dados

In [63]:
# a mesma função do coódigo anterior, porém ordenando os valores da coluna NM_UE por ordem alfabética
# para ordenar os dados, utilizamos o método sort_values()
total_cidades_pe['NM_UE'].sort_values().unique()

array(['ABREU E LIMA', 'AFOGADOS DA INGAZEIRA', 'AFRÂNIO', 'AGRESTINA',
       'ALAGOINHA', 'ALIANÇA', 'ALTINHO', 'AMARAJI', 'ANGELIM',
       'ARARIPINA', 'ARAÇOIABA', 'ARCOVERDE', 'BARRA DE GUABIRABA',
       'BARREIROS', 'BELO JARDIM', 'BELÉM DE MARIA',
       'BELÉM DO SÃO FRANCISCO', 'BETÂNIA', 'BEZERROS', 'BODOCÓ',
       'BOM CONSELHO', 'BOM JARDIM', 'BONITO', 'BREJINHO',
       'BREJO DA MADRE DE DEUS', 'BREJÃO', 'BUENOS AIRES', 'BUÍQUE',
       'CABO DE SANTO AGOSTINHO', 'CABROBÓ', 'CACHOEIRINHA', 'CAETÉS',
       'CALUMBI', 'CALÇADO', 'CAMARAGIBE', 'CAMOCIM DE SÃO FÉLIX',
       'CAMUTANGA', 'CANHOTINHO', 'CAPOEIRAS', 'CARNAUBEIRA DA PENHA',
       'CARNAÍBA', 'CARPINA', 'CARUARU', 'CASINHAS', 'CATENDE', 'CEDRO',
       'CHÃ DE ALEGRIA', 'CHÃ GRANDE', 'CONDADO', 'CORRENTES', 'CORTÊS',
       'CUMARU', 'CUPIRA', 'CUSTÓDIA', 'DORMENTES', 'ESCADA', 'EXU',
       'FEIRA NOVA', 'FERREIROS', 'FLORES', 'FLORESTA', 'FREI MIGUELINHO',
       'GAMELEIRA', 'GARANHUNS', 'GLÓRIA DO GOITÁ'

In [64]:
# verificando o tamanho da lista, o que significa, também, o número de cidades no Estado de Pernambuco
len(total_cidades_pe['NM_UE'].unique())

184

In [65]:
# o mesmo que o código anterior, porém utilizando outra variável
len(cidades_pe)

184

#### Encontrando os mesmos valores, porém utilizando set()

In [66]:
set(total_cidades_pe['NM_UE'])

{'ABREU E LIMA',
 'AFOGADOS DA INGAZEIRA',
 'AFRÂNIO',
 'AGRESTINA',
 'ALAGOINHA',
 'ALIANÇA',
 'ALTINHO',
 'AMARAJI',
 'ANGELIM',
 'ARARIPINA',
 'ARAÇOIABA',
 'ARCOVERDE',
 'BARRA DE GUABIRABA',
 'BARREIROS',
 'BELO JARDIM',
 'BELÉM DE MARIA',
 'BELÉM DO SÃO FRANCISCO',
 'BETÂNIA',
 'BEZERROS',
 'BODOCÓ',
 'BOM CONSELHO',
 'BOM JARDIM',
 'BONITO',
 'BREJINHO',
 'BREJO DA MADRE DE DEUS',
 'BREJÃO',
 'BUENOS AIRES',
 'BUÍQUE',
 'CABO DE SANTO AGOSTINHO',
 'CABROBÓ',
 'CACHOEIRINHA',
 'CAETÉS',
 'CALUMBI',
 'CALÇADO',
 'CAMARAGIBE',
 'CAMOCIM DE SÃO FÉLIX',
 'CAMUTANGA',
 'CANHOTINHO',
 'CAPOEIRAS',
 'CARNAUBEIRA DA PENHA',
 'CARNAÍBA',
 'CARPINA',
 'CARUARU',
 'CASINHAS',
 'CATENDE',
 'CEDRO',
 'CHÃ DE ALEGRIA',
 'CHÃ GRANDE',
 'CONDADO',
 'CORRENTES',
 'CORTÊS',
 'CUMARU',
 'CUPIRA',
 'CUSTÓDIA',
 'DORMENTES',
 'ESCADA',
 'EXU',
 'FEIRA NOVA',
 'FERREIROS',
 'FLORES',
 'FLORESTA',
 'FREI MIGUELINHO',
 'GAMELEIRA',
 'GARANHUNS',
 'GLÓRIA DO GOITÁ',
 'GOIANA',
 'GRANITO',
 'GRAVATÁ',
 'I

In [67]:
len(set(total_cidades_pe['NM_UE']))

184

In [68]:
# filtrando os dados relacionado ao Estado do Rio Grande do Norte
total_cidades_rn = df_completo[df_completo['SG_UF'] == 'RN']

In [69]:
# colocando os valores únicos de cidades de RN em uma lista
cidades_rn = total_cidades_rn['NM_UE'].unique()
cidades_rn

array(['MESSIAS TARGINO', 'TENENTE ANANIAS', 'PATU', 'CANGUARETAMA',
       'CAICÓ', 'BARAÚNA', 'TIBAU DO SUL', 'IELMO MARINHO', 'ACARI',
       'SÃO RAFAEL', 'CEARÁ-MIRIM', 'SERRA NEGRA DO NORTE',
       'LAGOA SALGADA', 'ITAJÁ', 'RIO DO FOGO', 'IPUEIRA', 'RIACHUELO',
       'AREIA BRANCA', 'CAIÇARA DO NORTE', 'JUCURUTU', 'ITAÚ', 'BREJINHO',
       'NATAL', 'BENTO FERNANDES', 'ANGICOS', 'MOSSORÓ', 'SANTO ANTÔNIO',
       'CRUZETA', 'PEDRA GRANDE', 'FRUTUOSO GOMES', 'CARNAÚBA DOS DANTAS',
       'LAGOA NOVA', 'SENADOR GEORGINO AVELINO', 'SERRA DE SÃO BENTO',
       'ALEXANDRIA', 'FRANCISCO DANTAS', 'RUY BARBOSA', 'PARNAMIRIM',
       'FERNANDO PEDROZA', 'SANTANA DO SERIDÓ', 'ALTO DO RODRIGUES',
       'GALINHOS', 'CERRO CORÁ', 'RIACHO DA CRUZ', 'FELIPE GUERRA',
       'MAXARANGUAPE', 'EXTREMOZ', 'SERRA CAIADA', 'POÇO BRANCO',
       'GROSSOS', 'MACAU', 'ASSÚ', 'PARELHAS', 'APODI', 'TOUROS',
       'SÃO JOSÉ DE MIPIBU', 'LUÍS GOMES', "LAGOA D'ANTA",
       'JARDIM DO SERIDÓ', 'MONTE ALE

In [70]:
# ordenando os dados únicos das cidades de RN por ordem alfabética
total_cidades_rn['NM_UE'].sort_values().unique()

array(['ACARI', 'AFONSO BEZERRA', 'ALEXANDRIA', 'ALMINO AFONSO',
       'ALTO DO RODRIGUES', 'ANGICOS', 'ANTÔNIO MARTINS', 'APODI',
       'AREIA BRANCA', 'AREZ', 'ASSÚ', 'BARAÚNA', 'BARCELONA',
       'BAÍA FORMOSA', 'BENTO FERNANDES', 'BOA SAÚDE', 'BODÓ',
       'BOM JESUS', 'BREJINHO', 'CAICÓ', 'CAIÇARA DO NORTE',
       'CAIÇARA DO RIO DO VENTO', 'CAMPO GRANDE', 'CAMPO REDONDO',
       'CANGUARETAMA', 'CARAÚBAS', 'CARNAUBAIS', 'CARNAÚBA DOS DANTAS',
       'CEARÁ-MIRIM', 'CERRO CORÁ', 'CORONEL EZEQUIEL',
       'CORONEL JOÃO PESSOA', 'CRUZETA', 'CURRAIS NOVOS',
       'DOUTOR SEVERIANO', 'ENCANTO', 'EQUADOR', 'ESPÍRITO SANTO',
       'EXTREMOZ', 'FELIPE GUERRA', 'FERNANDO PEDROZA', 'FLORÂNIA',
       'FRANCISCO DANTAS', 'FRUTUOSO GOMES', 'GALINHOS', 'GOIANINHA',
       'GOVERNADOR DIX-SEPT ROSADO', 'GROSSOS', 'GUAMARÉ',
       'IELMO MARINHO', 'IPANGUAÇU', 'IPUEIRA', 'ITAJÁ', 'ITAÚ',
       'JANDAÍRA', 'JANDUÍS', 'JAPI', 'JARDIM DE ANGICOS',
       'JARDIM DE PIRANHAS', 'JARDIM DO 

In [71]:
# verificando o total de cidades do Estado do RN
len(total_cidades_rn['NM_UE'].sort_values().unique())

167

In [72]:
# verificando o total de cidades do Estado do RN por meio de outra variável
len(cidades_rn)

167

#### Encontrando as mesmas informações, porém com o construtor set()

In [74]:
set(total_cidades_rn['NM_UE'])

{'ACARI',
 'AFONSO BEZERRA',
 'ALEXANDRIA',
 'ALMINO AFONSO',
 'ALTO DO RODRIGUES',
 'ANGICOS',
 'ANTÔNIO MARTINS',
 'APODI',
 'AREIA BRANCA',
 'AREZ',
 'ASSÚ',
 'BARAÚNA',
 'BARCELONA',
 'BAÍA FORMOSA',
 'BENTO FERNANDES',
 'BOA SAÚDE',
 'BODÓ',
 'BOM JESUS',
 'BREJINHO',
 'CAICÓ',
 'CAIÇARA DO NORTE',
 'CAIÇARA DO RIO DO VENTO',
 'CAMPO GRANDE',
 'CAMPO REDONDO',
 'CANGUARETAMA',
 'CARAÚBAS',
 'CARNAUBAIS',
 'CARNAÚBA DOS DANTAS',
 'CEARÁ-MIRIM',
 'CERRO CORÁ',
 'CORONEL EZEQUIEL',
 'CORONEL JOÃO PESSOA',
 'CRUZETA',
 'CURRAIS NOVOS',
 'DOUTOR SEVERIANO',
 'ENCANTO',
 'EQUADOR',
 'ESPÍRITO SANTO',
 'EXTREMOZ',
 'FELIPE GUERRA',
 'FERNANDO PEDROZA',
 'FLORÂNIA',
 'FRANCISCO DANTAS',
 'FRUTUOSO GOMES',
 'GALINHOS',
 'GOIANINHA',
 'GOVERNADOR DIX-SEPT ROSADO',
 'GROSSOS',
 'GUAMARÉ',
 'IELMO MARINHO',
 'IPANGUAÇU',
 'IPUEIRA',
 'ITAJÁ',
 'ITAÚ',
 'JANDAÍRA',
 'JANDUÍS',
 'JAPI',
 'JARDIM DE ANGICOS',
 'JARDIM DE PIRANHAS',
 'JARDIM DO SERIDÓ',
 'JAÇANÃ',
 'JOSÉ DA PENHA',
 'JOÃO CÂMARA'

In [75]:
len(set(total_cidades_rn['NM_UE']))

167

In [78]:
# transformando em inteiro o número de cidades de cada Estado
nr_cd_rn = int(len(cidades_rn))
nr_cd_pe = int(len(cidades_pe))

In [79]:
# somando o total de cidades de cada Estado
total_cidades = nr_cd_rn + nr_cd_pe
total_cidades

351

In [80]:
# aqui, verificamos através do método unique() a quantidade de cidades que o dataframe COMPLETO possui
df_total_cidades = len(df_completo['NM_UE'].unique())
df_total_cidades

348

### Como mostram as duas linhas acima, o número total de cidades se altera quando pesquisamos essa informação a partir de métodos diferentes. Ao filtrar os dados a partir dos Estados, temos um total de 351 cidades. Já quando analisamos o dataframe inteiro, sem filtrar os Estados inidividualmente, encontramos um total de 348 cidades. Isso significa que os dois Estados possuem cidades com o mesmo nome

### Para encontrar esses valores duplicados, vamos criar listas de valores únicos das cidades de cada Estado

In [81]:
# criando uma lista de valores únicos para as cidades do Estado de RN. Vamos deixar a lista em ordem alfabética
ue_rn = total_cidades_rn['NM_UE'].sort_values().unique()

In [82]:
# lista de valores únicos para as cidades do Estado de PE, também em ordem alfabética
ue_pe = total_cidades_pe['NM_UE'].sort_values().unique()

In [83]:
# criando uma variável do tipo lista para inserir todas as cidades de cada estado
todas_cidades = []

In [84]:
# criando um iterador for para adicionar cada uma das cidades de ue_rn na lista de todas_cidades
for cidade in ue_rn:
    todas_cidades.append(cidade)

In [85]:
# o mesmo para as cidades de PE. Assim, a lista todas_cidades vai possuir as cidades de ambos os Estado
for cidade in ue_pe:
    todas_cidades.append(cidade)

In [86]:
# conferindo o tamanho da lista todas_cidades
len(todas_cidades)

351

In [87]:
# uma das formas de verificarmos quais são as cidades repetidas é criar uma variavel do tipo lista, que, a partir da
# ação de um iterador for, irá receber os valores que aparecerem mais de uma vez na lista todas_cidades;
# em outras palavras, o iterador irá contar, através do método count(), quantas vezes determinado valor aparece na lista.
# Caso o valor apareça mais de uma vez, ele será inserido na variável lista cidade[]; assim, saberemos quais nomes de cidades
# existem em ambos os Estados
cidade = []
for x in todas_cidades:
    if todas_cidades.count(x) > 1:
        cidade.append(x)
cidade

['BREJINHO',
 'PARNAMIRIM',
 'SANTA CRUZ',
 'BREJINHO',
 'PARNAMIRIM',
 'SANTA CRUZ']

In [88]:
# outra forma, mais simples, de encontrar os valores repetidos, é utilizar o método value_counts()
# o método value_counts(), porém, só existe EM DATAFRAMES OU SEMELHANTES, COMO SÉRIES, E NÃO EM LISTAS
# assim, para que possamos utilizar value_counts() em todas_cidades, temos, primeiro, que converter a variável do tipo
# lista em objeto Série. Assim, podemos criar uma variável para receber a Série todas_cidades e contar os valores
df_cidades = pd.Series(todas_cidades).value_counts()
df_cidades
# podemos ver como os valores duplicados (ou maiores que um, se fosse o caso) também podem ser encontrados de outra forma

BREJINHO             2
PARNAMIRIM           2
SANTA CRUZ           2
ACARI                1
GOIANA               1
                    ..
RODOLFO FERNANDES    1
RIO DO FOGO          1
RIACHUELO            1
RIACHO DE SANTANA    1
ÁGUAS BELAS          1
Length: 348, dtype: int64

In [92]:
# podemos filtrar a série para verificar quais valores são maiores que 1 (ou seja, que existem em duplicidade na Série)
df_cidades[df_cidades > 1]

BREJINHO      2
PARNAMIRIM    2
SANTA CRUZ    2
dtype: int64

In [93]:
name = df_cidades.index[:3]
name

Index(['BREJINHO', 'PARNAMIRIM', 'SANTA CRUZ'], dtype='object')

#### Podemos, agora, encontrar o número de candidaturas para cada uma das cidades repetidas, para praticar aquilo que aprendemos anteriormente

In [95]:
# vamos colocar as três cidades em uma lista
cidades_duplicadas = []
cidades_duplicadas.append(cidade[0])
cidades_duplicadas.append(cidade[1])
cidades_duplicadas.append(cidade[2])
cidades_duplicadas

['BREJINHO', 'PARNAMIRIM', 'SANTA CRUZ']

#### Total de candidaturas da cidade de Brejinho - PE

In [96]:
cand_brejinho_pe = df_completo[(df_completo['SG_UF'] == 'PE') & (df_completo['NM_UE'] == 'BREJINHO')]
cand_brejinho_pe

DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
685    05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
1448   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
1556   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
2724   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
3189   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
3439   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
3691   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
3748   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
3858   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
6190   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
6434   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
6906   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
9657   05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
10122  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
10175  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
10244  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
11339  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
12445  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
12972  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
13484  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
13617  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
14247  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
14627  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
15173  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
16399  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
17255  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
17437  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
18189  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
19662  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
19938  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
20660  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
20857  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   

       NR_TURNO  CD_ELEICAO                DS_ELEICAO  DT_ELEICAO  \
685           1         426  Eleições Municipais 2020  15/11/2020   
1448          1         426  Eleições Municipais 2020  15/11/2020   
1556          1         426  Eleições Municipais 2020  15/11/2020   
2724          1         426  Eleições Municipais 2020  15/11/2020   
3189          1         426  Eleições Municipais 2020  15/11/2020   
3439          1         426  Eleições Municipais 2020  15/11/2020   
3691          1         426  Eleições Municipais 2020  15/11/2020   
3748          1         426  Eleições Municipais 2020  15/11/2020   
3858          1         426  Eleições Municipais 2020  15/11/2020   
6190          1         426  Eleições Municipais 2020  15/11/2020   
6434          1         426  Eleições Municipais 2020  15/11/2020   
6906          1         426  Eleições Municipais 2020  15/11/2020   
9657          1         426  Eleições Municipais 2020  15/11/2020   
10122         1         426  Eleições Municipais 2020  15/11/2020   
10175         1         426  Eleições Municipais 2020  15/11/2020   
10244         1         426  Eleições Municipais 2020  15/11/2020   
11339         1         426  Eleições Municipais 2020  15/11/2020   
12445         1         426  Eleições Municipais 2020  15/11/2020   
12972         1       

In [97]:
cand_brejinho_pe.shape

(32, 63)

In [105]:
candidaturas_pe[candidaturas_pe['NM_UE'] == 'BREJINHO'].shape

(32, 63)

#### Total de candidaturas da cidade de Brejinho - RN

In [100]:
cand_brejinho_rn = df_completo[(df_completo['SG_UF'] == 'RN') & (df_completo['NM_UE'] == 'BREJINHO')]
cand_brejinho_rn

DT_GERACAO HH_GERACAO  ANO_ELEICAO  CD_TIPO_ELEICAO    NM_TIPO_ELEICAO  \
21123  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21245  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21271  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21710  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
21953  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
22578  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
22905  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23075  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23172  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23217  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23292  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23354  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23608  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23619  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
23732  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
24025  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
24643  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
24764  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
24899  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
24979  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
25420  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
25620  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
25659  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
25786  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
25831  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
25955  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
26120  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
26131  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
26282  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
26365  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
26587  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
26738  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27141  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27169  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27314  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27417  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27501  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27558  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27762  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
27906  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
28121  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
28145  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
28595  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
28612  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
28836  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
29041  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
29052  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
29094  05/08/2021   12:18:13         2020                2  ELEIÇÃO ORDINÁRIA   
29387  05/08/2021   12:18:13         2

In [127]:
rn = dict(df_RN['NM_UE'].value_counts())
rn

{'NATAL': 766,
 'MOSSORÓ': 488,
 'PARNAMIRIM': 351,
 'CANGUARETAMA': 214,
 'EXTREMOZ': 213,
 'CAICÓ': 212,
 'CURRAIS NOVOS': 211,
 'MACAÍBA': 198,
 'CEARÁ-MIRIM': 190,
 'SÃO GONÇALO DO AMARANTE': 183,
 'TOUROS': 182,
 'MACAU': 181,
 'ASSÚ': 161,
 'BARAÚNA': 141,
 'SÃO JOSÉ DE MIPIBU': 129,
 'GOVERNADOR DIX-SEPT ROSADO': 111,
 'AREIA BRANCA': 106,
 'IELMO MARINHO': 98,
 'APODI': 95,
 'NOVA CRUZ': 94,
 'PEDRO VELHO': 91,
 'SANTA CRUZ': 89,
 'GROSSOS': 85,
 'BAÍA FORMOSA': 85,
 'TIBAU DO SUL': 84,
 'NÍSIA FLORESTA': 82,
 'GOIANINHA': 81,
 'JOÃO CÂMARA': 78,
 'POÇO BRANCO': 78,
 'PENDÊNCIAS': 76,
 'RIO DO FOGO': 75,
 'PAU DOS FERROS': 72,
 'IPANGUAÇU': 72,
 'SÃO PAULO DO POTENGI': 70,
 'SANTA MARIA': 69,
 'LAJES': 68,
 'PUREZA': 67,
 'TANGARÁ': 67,
 'LAGOA NOVA': 62,
 'SANTO ANTÔNIO': 62,
 'SÃO MIGUEL DO GOSTOSO': 62,
 'TAIPU': 61,
 'GUAMARÉ': 60,
 'BOA SAÚDE': 60,
 'MAXARANGUAPE': 59,
 'MONTANHAS': 59,
 'JUCURUTU': 59,
 'AREZ': 58,
 'JARDIM DE PIRANHAS': 57,
 'SÃO MIGUEL': 57,
 'SENADOR E

In [101]:
cand_brejinho_rn.shape

(52, 63)

#### Tentando encontrar o número de candidaturas de todas as cidades repetidas no dataframe, por Estado

In [122]:
txts = []
for uf in lista_uf:
    for cidade in cidades_duplicadas:
        candidaturas = df_completo[(df_completo['SG_UF'] == uf) & (df_completo['NM_UE'] == cidade)].shape[0]
        txt = f'O total de candidaturas na cidade de {cidade} - {uf} foi de {candidaturas}'
        txts.append(txt)

In [123]:
txts

['O total de candidaturas na cidade de BREJINHO - PE foi de 32',
 'O total de candidaturas na cidade de PARNAMIRIM - PE foi de 35',
 'O total de candidaturas na cidade de SANTA CRUZ - PE foi de 27',
 'O total de candidaturas na cidade de BREJINHO - RN foi de 52',
 'O total de candidaturas na cidade de PARNAMIRIM - RN foi de 351',
 'O total de candidaturas na cidade de SANTA CRUZ - RN foi de 89']